### PR TITLE
DAOS-11613 test: reproducer for multi EQ / multi thread issue

### PR DESCRIPTION
**Pull request by mchaarawi**
Required-githooks: true

Signed-off-by: Mohamad Chaarawi <mohamad.chaarawi@intel.com>

### CODE REVIEW

**Review by wangshilong — APPROVED** [review]
Just wondering, we used different eq for fio dfs as well, i did not hit this issue when running fio multiple threads testing..

**Comment by mchaarawi** [comment]
> Just wondering, we used different eq for fio dfs as well, i did not hit this issue when running fio multiple threads testing..

with FIO we use 1 EQ per process, not per thread i think?

**Comment by wangshilong** [comment]
> > Just wondering, we used different eq for fio dfs as well, i did not hit this issue when running fio multiple threads testing..
> 
> with FIO we use 1 EQ per process, not per thread i think?

daos_fio_init is called per thread? (i mean in multiple thread case), it call daos_eq_create

**Comment by wangshilong** [comment]
> > > > Just wondering, we used different eq for fio dfs as well, i did not hit this issue when running fio multiple threads testing..
> > > 
> > > 
> > > with FIO we use 1 EQ per process, not per thread i think?
> > 
> > 
> > daos_fio_init is called per thread? (i mean in multiple thread case), it call daos_eq_create
> 
> indeed. you are right. however you won't see this in FIO because that init is protected by mutex. so the context creation from multiple threads is serialized and does not happen concurrently, which is the issue that was fixed in Alex's PR that landed recently.

indeed...i missed mutex there..

**Review by jolivier23 — APPROVED** [review]
So the issue has been fixed?

**Comment by mchaarawi** [comment]
> So the issue has been fixed?

yes. test now passes

File: src/tests/suite/daos_obj_array.c [hunk 1]
```diff
@@ -1397,6 +1397,141 @@ cond_ops(void **state)
...
+{
+	struct test_async_arg	*targ = arg;
+	d_iov_t			dkey[NUM_IOS];
+	d_sg_list_t		sgls[NUM_IOS];
+	d_iov_t			sg_iovs[NUM_IOS];
+	daos_iod_t		iods[NUM_IOS];
+	daos_recx_t		recx[NUM_IOS];
+	char			akey_str[NUM_IOS][10];
+	char			dkey_str[NUM_IOS][10];
+	char			buf[NUM_IOS][IO_SIZE];
```
Comment (frostedcmos) [inline src/tests/suite/daos_obj_array.c:1426]: with num_ios set to 256 thats a lot of stuff on a stack. does it survive without going over the limits?

**Review by frostedcmos — APPROVED** [review]

File: src/tests/suite/daos_obj_array.c (hunk 1 above)
Comment (mchaarawi) [reply src/tests/suite/daos_obj_array.c:1426]: IO size is pretty small though

Comment (jolivier23) [reply src/tests/suite/daos_obj_array.c:1426]: In this case, it's a normal pthread rather than a ULT.  Google tells me default in 96K

Comment (jolivier23) [reply src/tests/suite/daos_obj_array.c:1426]: FWIW, scons install -j49 --analyze-stack="-i daos_obj_array" shows this for that function stack frame size.

50376 bytes:
    src/tests/suite/daos_obj_array.c:1416:test_async_io

**Pull request merged**